Source/Details/ASPageTable.mm

```diff
@@ -0,0 +1,149 @@
+//
+//  ASPageTable.mm
+//  Texture
+//
+//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <AsyncDisplayKit/ASPageTable.h>
+
+#import <AsyncDisplayKit/ASEqualityHashHelpers.h>
+#import <AsyncDisplayKit/ASEqualityHelpers.h>
+
+@implementation ASPageCoordinates
+
+- (instancetype)initWithX:(NSUInteger)x Y:(NSUInteger)y
+{
+  self = [super init];
+  if (self) {
+    _x = x;
+    _y = y;
+  }
+  return self;
+}
+
++ (instancetype)pageCoordinatesWithX:(NSUInteger)x Y:(NSUInteger)y
+{
+  return [[ASPageCoordinates alloc] initWithX:x Y:y];
+}
+
++ (instancetype)pageCoordinatesForPageThatContainsPoint:(CGPoint)point pageSize:(CGSize)pageSize
+{
+  return [[ASPageCoordinates alloc] initWithX:(MAX(0.0f, point.x) / pageSize.width) Y:(MAX(0.0f, point.y) / pageSize.height)];
+}
+
++ (NSArray<ASPageCoordinates *> *)pageCoordinatesForPagesInRect:(CGRect)rect withContentSize:(CGSize)contentSize pageSize:(CGSize)pageSize
+{
+  CGRect contentRect = CGRectMake(0.0f, 0.0f, contentSize.width, contentSize.height);
```

Comment:
Let's never use `3.0f` `0.0f` etc – the compiler will choose the appropriate type. CGFloat is double on 64-bit, after all.